src/tests.py

```diff
@@ -565,5 +620,140 @@ def before_call(self, cb, func, *args, **kwargs):
+
+        self._mock_function(self.breaker, _inc_counter)
+        self._start_threads(trigger_error, 3)
+        self.assertEqual(1500, self.breaker.fail_counter)
+
+    def test_success_thread_safety(self):
+        """CircuitBreaker: it should compute a successful call atomically
+        to avoid race conditions.
+        """
+        @self.breaker
+        def suc(): return True
+
+        def trigger_success():
+            for n in range(500):
+                suc()
+
+        class SuccessListener(CircuitBreakerListener):
+            def success(self, cb):
+                c = 0
+                if hasattr(cb, '_success_counter'):
+                    c = cb._success_counter
+                sleep(0.00005)
+                cb._success_counter = c + 1
+
+        self.breaker.add_listener(SuccessListener())
+        self._start_threads(trigger_success, 3)
+        self.assertEqual(1500, self.breaker._success_counter)
+
+    def test_half_open_thread_safety(self):
+        """CircuitBreaker: it should allow only one trial call when the
+        circuit is half-open.
+        """
+        self.breaker = CircuitBreaker(fail_max=1, reset_timeout=0.01)
+
+        self.breaker.open()
+        sleep(0.01)
+
+        @self.breaker
+        def err(): raise Exception()
+
+        def trigger_failure():
+            try: err()
+            except: pass
+
+        class StateListener(CircuitBreakerListener):
+            def __init__(self):
+                self._count = 0
+
+            def before_call(self, cb, fun, *args, **kwargs):
+                sleep(0.00005)
+
+            def state_change(self, cb, old_state, new_state):
+                if new_state.name == 'half-open':
+                    self._count += 1
+
+        state_listener = StateListener()
+        self.breaker.add_listener(state_listener)
+
+        self._start_threads(trigger_failure, 5)
+        self.assertEqual(1, state_listener._count)
+
+
+    def test_fail_max_thread_safety(self):
+        """CircuitBreaker: it should not allow more failed calls than 'fail_max'
+        setting. Note that with Redis, where we have separate systems
+        incrementing the counter, we can get concurrent updates such that the
+        counter is greater than the 'fail_max' by the number of systems. To
+        prevent this, we'd need to take out a lock amongst all systems before
+        trying the call.
+        """
+        @self.breaker
+        def err(): raise Exception()
+
+        def trigger_error():
+            for i in range(2000):
+                try: err()
+                except: pass
+
+        class SleepListener(CircuitBreakerListener):
+            def before_call(self, cb, func, *args, **kwargs):
+                sleep(0.00005)
+
+        self.breaker.add_listener(SleepListener())
+        num_threads = 3
+        self._start_threads(trigger_error, num_threads)
+        self.assertTrue(self.breaker.fail_counter < self.breaker.fail_max + num_threads)
```

Comment:
@phillbaker sorry for the delay!

I don't think we need to prevent this situation; IMO it's better to allow a few more calls go through before opening the circuit than locking around every call.